src/Service/SigningService.php

```diff
@@ -0,0 +1,362 @@
+   * User has returned from a signing service, clearing cookies, redirecting
+   * to the forward url.
+   *
+   * @params string $file
+   *   File name.
+   * @oaram string $action
+   *   Expected action is 'result' or 'cancel'.
+   *
+   * @return TrustedRedirectResponse
+   *   Redirect response.
+   */
+  public function result(string $file, string $action) : TrustedRedirectResponse {
+    $request = $this->requestStack->getCurrentRequest();
+
+    if (empty($cookie = $request->cookies->get(SigningService::FORWARD_URL_COOKIE) ?? NULL)) {
+      throw new AccessDeniedHttpException('Required cookie not found.');
+    }
+
+    $url = base64_decode($cookie);
+    if (!$this->isValidUrl($url)) {
+      throw new AccessDeniedHttpException("Provided cookie forward url value not accepted: $url");
+    }
+
+    $sep = str_contains($url, '?') ? '&' : '?';
+    $response = new TrustedRedirectResponse($url . $sep . http_build_query(['file' => $file, 'action' => $action]));
+
+    // Deleting cookies.
+    $response->headers->clearCookie(SigningService::FORWARD_URL_COOKIE);
+
+    return $response;
+  }
+
+  /**
+   * Download the signed file.
+   */
+
+  /**
+   * Sends the file as a binary response.
+   *
+   * @param string $file
+   *   File to attempt download for.
+   * @param bool $leave
+   *   If a file should be left on the server.
+   *   TRUE - file remains untouched.
+   *   FALSE - file will be deleted after sending.
+   * @return BinaryFileResponse
+   *   Found a file.
+   */
+  public function download(string $file, bool $leave) : BinaryFileResponse {
+    if (!preg_match('/^[a-z0-9]{32}\.pdf$/', $file)) {
+      throw new BadRequestHttpException("Invalid file name: $file. Must be contain letters or numbers and be 32 chars long");
+    }
+
+    $signedPdf = $this->getSignedFilesDir() . '/' . substr($file, 0, 32) . '-signed.pdf';
+
+    $filesize = (int) @filesize($signedPdf);
+    $this->debug('Sending %signed_pdf (%filesize bytes)', ['%signed_pdf' => $signedPdf, '%filesize' => $filesize]);
+    if ($filesize) {
+      $response = new BinaryFileResponse($signedPdf);
+      $response->headers->set('Content-Type', 'application/pdf');
+      $response->headers->set('Content-Length', $filesize);
+
+      // Unless told otherwise, remove the file after sending.
+      $deleteFileAfterSend = !$leave;
+      $response->deleteFileAfterSend($deleteFileAfterSend);
+
+      return $response;
+    }
+    else {
+      $this->debug('Attempt to download empty or nonexisting file %file', ['%file' => $file]);
+      throw new AccessDeniedHttpException('Attempt to download empty or nonexisting file');
+    }
+  }
+
+  /**
+   * Checks the URL.
+   *
+   * Checks if URL belongs to the allowed domains.
+   *
+   * @param string $url
+   *   The url to check.
+   *
+   * @return bool
+   *   TRUE if url is known, otherwise FALSE.
+   */
+  private function isValidUrl(string $url) : bool {
+    $prefix = preg_replace('@https?://([^/:]*).*@', '$1', $url);
```

Comment:
```suggestion
    $host = parse_url($url, PHP_URL_HOST),
```
(cf. https://www.php.net/manual/en/function.parse-url.php).